runtime/lua/vim/_extui/messages.lua

```diff
@@ -472,16 +473,37 @@ function M.set_pos(type)
       -- Ensure last line is visible and first line is at top of window.
       local row = (texth.all > height and texth.end_row or 0) + 1
       api.nvim_win_set_cursor(ext.wins.msg, { row, 0 })
-    elseif type == 'pager' and api.nvim_win_get_config(win).hide then
-      -- Cannot leave the cmdwin to enter the pager, so close it.
-      -- NOTE: regression w.r.t. the message grid, which allowed this. Resolving
-      -- that would require somehow bypassing textlock for the pager.
+    elseif type == 'pager' then
       if fn.getcmdwintype() ~= '' then
-        api.nvim_command('quit')
+        if will_pager then
+          config.relative, config.win, config.row, config.col = 'win', 0, 0, 0
+        else
+          -- Cannot leave the cmdwin to enter the pager, so close it.
+          -- NOTE: regression w.r.t. the message grid, which allowed this.
+          -- Resolving that would require somehow bypassing textlock for the pager.
+          api.nvim_command('quit')
+        end
       end
 
+      routed = will_pager
       -- It's actually closed one event iteration later so schedule in case it was open.
       vim.schedule(function()
+        -- For a routed message close the pager on the next typed mapping, unless entered.
```

Comment:
```suggestion
        -- For a routed message close the pager on any user input, unless the pager is focused.
```

Reply:
I noticed the vim.on_key() callback doesn't trigger until a complete mapping, or 'timeoutlen'. So it's not exactly any user input right?

Reply:
Yes sorry, unmapped keys, complete mapping or passed timeoutlen. Point was that it waits for a potential mapping, which is actually useful here to automatically support /<C-w>-w.